### PR TITLE
[RFC] mtest: use asyncio instead of concurrency.futures

### PR DESCRIPTION
**Pull request by bonzini**
This pull request rewrites the SingleTestRunner and TestHarness to use asyncio for subprocesses.

The main advantage is better control on cancellation; with the current code, KeyboardInterrupt was never handled by the thread executor so the code that tried to handle it in SingleTestRunner only worked for non-parallel tests.  Likewise, SIGTERM was not handled and left tests running. (In addition, if the test exited with exit status 0 in response to SIGTERM, it was not marked as a failure).

With this pull request, mtest gains a new `INTERRUPT` TestResult and SIGINT will return it for the oldest-running test. SIGTERM instead will return it for all running tests and will not start any new test. Because a picture is worth a thousand works, here are some Asciinema recordings:
* [sending SIGINT to `meson test` (unpatched)](https://asciinema.org/a/QdFqzfET6mhxZk6BcokKo1pDx)
* [sending SIGINT to `meson test` (patched)](https://asciinema.org/a/TwPwquCNiZFtfB7SjOzrAm0K1)
* [sending SIGTERM to `meson test` (patched)](https://asciinema.org/a/WsK001eLfvPueEzz7PESSDq2G)

The pull request is split between a refactoring part and a new feature part, so it can be split easily if desired.

### CODE REVIEW

**Comment by bonzini** [comment]
(According to CI there seems to be a race that I cannot figure out. Not a big deal for a generic review of the idea).

**Comment by bonzini** [comment]
It seems like mtest occasionally never exits on Window I'll try to bisect to find what is wrong.

**Comment by nirbheek** [comment]
You are my hero.

**Comment by bonzini** [comment]
So it seems that asyncio.subprocess was the result of the breakage on Windows. Trying again without.

**Comment by nirbheek** [comment]
> It seems like mtest occasionally never exits on Window I'll try to bisect to find what is wrong.

I ran the code in mtest-asyncio-full on the `test_suite_selection` unit test for 100 iterations and it didn't hang even once. Very strange that it hung on the CI.

File: mesonbuild/mtest.py
Comment (dcbaker) [inline mesonbuild/mtest.py:722]: why are you using communicate if you're not getting back the stdout or stderr? Wouldn't it be better to use `Popen.wait`?

Comment (bonzini) [reply mesonbuild/mtest.py:722]: The coffee that was there before also used `communicate`, but yeah `wait` is better.

**Comment by bonzini** [comment]
I don't have any more parts of the patches that I can bisect, but it looks like ProactorEventLoop is responsible for the issue. I'll wait for [this CI run](https://dev.azure.com/jussi0947/jussi/_build/results?buildId=8072) to finish, and then trigger another without it.

OTOH this failing run has Python 3.5.4, therefore the issue with ProactorEventLoop may fix itself once Windows CI is updated to 3.6.

**Comment by scivision** [comment]
This is a side comment--bonzini is far more experienced with this than me.

My experience is with ~ 1000 users, mostly Windows, Python >= 3.6..3.8, and that ProactorEventLoop and subprocess worked well for that user group.  In my programs, I explicitly selected ProactorEventLoop for Windows via `if` statement.

### Question

Is it worth having (do we already have?) a Windows CI on Python 3.6, and another Windows CI on Python 3.9, since **Python 3.8 changed Windows [default event loop to Proactor](https://docs.python.org/3/library/asyncio-policy.html#asyncio.DefaultEventLoopPolicy)**. Python 3.6 and 3.7 [default event loop is Selector on Windows](https://docs.python.org/3.7/library/asyncio-policy.html#asyncio.DefaultEventLoopPolicy)

**Comment by bonzini** [comment]
We would use Proactor anyway once we switch to a new-enough Python that fixed the bugs I was encountering. But yeah Windows should have one CI on the oldest possible Python version and one on a newer version.

File: mesonbuild/mtest.py
Comment (dcbaker) [inline mesonbuild/mtest.py:726]: I think I know why you're catching `BaseException`, but since this is normally a bad thing, could you add a comment explaining why you're catching it.

**Review by dcbaker — APPROVED** [review]
I have on small nit, but otherwise this looks great. I'll be happy to merge this as soon as we get 0.56.0 out the door, and the bump to python 3.6 happens (which should be right after 0.57 development starts).

File: mesonbuild/mtest.py
Comment (bonzini) [reply mesonbuild/mtest.py:726]: Sure. Note that if the CI passes I wouldn't mind going straight to the version that uses asyncio subprocesses, on which case the catching of BaseException disappears.

Comment (dcbaker) [reply mesonbuild/mtest.py:726]: I'd really prefer to use asyncio.subprocess if it will work, just because it's going to be less maintenance for us. If the performance is worse, maybe we can consider using a faster event loop in CI, like the libuv one.

**Comment by jpakkane** [comment]
Python 3.5 is dead in master now.

**Comment by bonzini** [comment]
Looks like 3.6 is okay with SelectorEventLoop, I will add the commits that switch to Proctor once the run is over.

**Comment by jpakkane** [comment]
Since I am paranoid and the CI system has been revamped, please rebase so the new CI runs this as well. Thanks.

**Comment by bonzini** [comment]
Not sure why `test cases/d/9 features` is failing but it's the same on several of my other pull requests (sometimes for `unity=off` and sometimes for `unity=on`).

**Comment by dcbaker** [comment]
I'm seeing that too. Macos just seems generally unreliable unfortunately

**Comment by bonzini** [comment]
FWIW I have the next part ready which is to read logs via asyncio pipes instead of temporary files, and parsing TAP asynchronously (instead of waiting until the subprocess terminates). That alone doesn't add much value, but it is an important step towards reporting TAP results one by one; doing that for multiple concurrent tests would be almost impossible without asyncio. I might open a draft PR as soon as this one is merged, so as to get reviews.

**Pull request merged**
